lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -1901,10 +1902,60 @@ private function doMerge($entity, array &$visited, $prevManagedCopy = null, $ass
 
         $this->cascadeMerge($entity, $managedCopy, $visited);
 
+        // Restore previous state (where the given entity was previously managed)
+        if (!$isKnown) {
```

Comment:
This may should work as if the return managed entity is different from the input one we need to ensure the spl hash of the input entity is not used anymore as it could be garbage collected !